include/soci/type-holder.h

```diff
@@ -44,49 +50,310 @@ T* checked_ptr_cast(U* ptr)
     return static_cast<T*>(ptr);
 }
 
-// Base class holder + derived class type_holder for storing type data
-// instances in a container of holder objects
-template <typename T>
-class type_holder;
-
-class holder
+// Type safe conversion that fails at compilation if instantiated
```

Comment:
Is this really correct? I don't see why would it fail at compile-time, it seems to be failing at run-time instead.

Reply:
That's indeed not true. It's a leftover from the PR I've based this work on.